packages/phone-number-privacy/combiner/src/match-making/get-contact-matches.ts

```diff
@@ -36,16 +47,133 @@ export async function handleGetContactMatches(
+          // but we were unable to find a record of their previous phone number signature due to a db error.
+          logger.info(
+            { account },
+            'Allowing account to perform matchmaking due to db error finding phone number record. We will not record their phone number this time.'
+          )
+          matchmakingId = undefined
+        })
+        if (matchmakingId) {
+          if (!signedUserPhoneNumberRecord) {
+            // Account has performed matchmaking before and has provided a valid phone number dek signature
+            // but we do not have a record of their previous phone number signature in the db.
+            // This could be bc we were unable to verify the user phone number signature last time, or bc
+            // the user never provided a phone number signature (backwards compatibility).
+            logger.info(
+              { account },
+              'Allowing account to perform matchmaking since we have no record of the phone number it used before. We will record the phone number this time.'
+            )
+          } else {
+            if (signedUserPhoneNumberRecord !== matchmakingId.signedUserPhoneNumber) {
+              // Account has performed matchmaking before and has provided a valid phone number dek signature
+              // but the phone number signature we have stored in the db does not match what they provided.
+              const dekSignerRecord = await getDekSignerRecord(account, logger)
+              if (
+                dekSignerRecord &&
+                matchmakingId.dekSigner !== dekSignerRecord &&
+                verifySignature(userPhoneNumber, signedUserPhoneNumberRecord, dekSignerRecord)
+              ) {
+                // The phone number signature we have stored in the db wont match if the user has rotated their dek.
+                logger.info(
+                  {
+                    account,
+                    dekSignerRecord,
+                    dekSigner: matchmakingId.dekSigner,
+                  },
+                  'User has rotated their dek since first requesting matches. Updating dekSigner record for account.'
+                )
+              } else {
+                logger.info(
+                  { account },
+                  'Blocking account from querying matches for a different phone number than before.'
+                )
+                respondWithError(
+                  response,
+                  403,
+                  WarningMessage.DUPLICATE_REQUEST_TO_MATCHMAKE,
+                  logger
+                )
+                return
+              }
+            } else {
+              // Account has performed matchmaking before and has provided a phone number dek signature
+              // matching what we have stored in the db.
+              logger.info(
+                { account },
+                'Allowing account to requery matches for the same phone number as before.'
```

Comment:
It's not like it is any big problem, but generally I would 100% avoid doing comments like that trying to make sense out of the code. Maybe consider refactoring parts of the code into smaller functions, that have name, single purpose and are descriptive. This way we have less to maintain (comments tend to not be maintained and creates more problems than solutions), we also have smaller parts of the code that are easier to comprehend by others. (I know you know the code).

https://docs.codeclimate.com/docs/cognitive-complexity
Code complexity kills it, even if it is 100% correct. One year from now this will be much harder to comprehend.  (like I said, if you  feel the code is and will be ok then leave it. I accept the PR)

Reply:
@alecps  🥇 Nice 🥇 , looks much better and is easier to understand IMO. Even for me!